fdb-record-layer-spatial/src/main/java/com/geophile/z/async/SpatialJoinAsync.java

```diff
@@ -0,0 +1,202 @@
+        return new SpatialJoinImplAsync<>(duplicates, filter, null, null);
+    }
+
+    /**
+     * Creates and configures a new SpatialJoin object.
+     *
+     * @param duplicates Indicates whether the spatial join will suppress duplicates.
+     * @param <LEFT>     Type of object for left side of join
+     * @param <RIGHT>    Type of object for right side of join
+     * @return A configured SpatialJoin object. All spatial joins computed using it will use the configuration
+     * specified by the above arguments. False positives will be included in spatial join output.
+     */
+    public static <LEFT, RIGHT> SpatialJoinAsync<LEFT, RIGHT> newSpatialJoin(Duplicates duplicates)
+    {
+        return new SpatialJoinImplAsync<>(duplicates, null, null, null);
+    }
+
+    /**
+     * Returns an {@link java.util.Iterator} that will provide access to spatial join results.
+     * The objects accessed through the {@link java.util.Iterator} are {@link com.geophile.z.Pair}s,
+     * such that the left object comes from the leftSpatialIndex, and the right object comes from the rightSpatialIndex.
+     * The results are not filtered, and so may contain false positives.
+     *
+     * @param <LEFT_RECORD>     Type of {@link com.geophile.z.Record} in leftSpatialIndex.
+     * @param <RIGHT_RECORD>    Type of {@link com.geophile.z.Record} in rightSpatialIndex.
+     * @param leftSpatialIndex  One spatial join input.
+     * @param rightSpatialIndex The other spatial join input.
+     * @return An {@link java.util.Iterator} providing access to spatial join results.
+     */
+    public abstract <LEFT_RECORD extends Record, RIGHT_RECORD extends Record> IteratorAsync<Pair<LEFT_RECORD, RIGHT_RECORD>> iterator(SpatialIndexAsync<LEFT_RECORD> leftSpatialIndex,
+                                                                                                                                      SpatialIndexAsync<RIGHT_RECORD> rightSpatialIndex);
+
+    /**
+     * Returns an {@link java.util.Iterator} that will provide access to spatial join results.
+     * The objects accessed through the {@link java.util.Iterator} are {@link com.geophile.z.SpatialObject}s
+     * from the data argument that overlap the given query object.
+     * The results are filtered using the given filter, and should not contain false positives.
+     *
+     * @param <RECORD> Type of {@link com.geophile.z.Record} in data.
+     * @param query    Used to locate data elements of interest.
+     * @param data     The set of {@link com.geophile.z.SpatialObject}s to be searched.
+     * @return An {@link java.util.Iterator} providing access to spatial join results.
+     */
+    public abstract <RECORD extends Record> IteratorAsync<RECORD> iterator(SpatialObject query,
+                                                                           SpatialIndexAsync<RECORD> data);
+
+    /**
+     * Specifies duplicate-handling behavior for spatial joins.
+     */
+    public enum Duplicates
```

Comment:
It kind of feels like `SpatialJoin` and `SpatialJoinAsync` should use the same classes for `Duplicates`, `InputObserver`, and `Filter`, as I don't think they're any different. (I could see someone, someday wanting a `FilterAsync` and maybe an `InputObserverAsync`, I suppose.) I'm not sure how they're actually used, though, and if this API is all internal and can be changed, I suppose it doesn't matter too much at the moment.